Vending step8

AppleCEO opened the pull request:
요구사항 반영했습니다.

- 코어 그래픽스(Core Graphics) 를 이용해서 재고 현황을 관리자 뷰에 pie chart 로 그려줬습니다.

AppleCEO committed:
수정 - 파일명에 공백 때문에 이미지가 안뜨기에 이름 변경

AppleCEO committed:
추가 - 방향을 잘못잡고 cocoa pods 설치

AppleCEO committed:
추가 - UIView 추가

AppleCEO committed:
생성 - PieGraphView 생성

AppleCEO committed:
추가 - pieGraphDrawable 프로토콜 생성 및 구현

AppleCEO committed:
추가 - 음료 종류에 따라 이름을 한글로 리턴하는 메소드 추가

AppleCEO committed:
추가 - 결과화면 추가

godrm commented on VendingMachineApp/VendingMachineApp/AdminViewController.swift:
```diff
@@ -20,6 +21,9 @@ class AdminViewController: UIViewController {
     }
     
+    override func viewWillAppear(_ animated: Bool) {
+        drawPieGraph(vendingMachine)
```
세 가지 측면이 있습니다. 
- viewWillAppear() 같은 override 하는 메소드는 반드시 super.xxxx()를 추가하세요. 그렇지 않으면 오동작을 할 수 있습니다.
- 내부 메소드를 호출하는 데 왜 매개변수로 vendingMachine을 넘기는 건가요?
- 명시적으로 뷰를 draw하는 메소드를 호출하면 안됩니다. 뷰 드로잉 사이클 관련 문서를 찾아보세요. 다시 그리도록 지정하는 방법이 있습니다.

AppleCEO replied:
피드백 감사합니다.
- super.viewWillAppear(false) 메소드를 추가하였습니다.
- 의미없이 반복되면 vendingMachine 인스턴스를 매개변수로 넘기던 부분 수정하였습니다.
- draw 을 호출하고 있지는 않습니다. 
drawPieGraph 메소드에서는 데이터만 바꿔주고 데이터가 바뀌면 setNeedsDisplay 가 호출되어서 그림이 그려지고 있습니다. 
메소드명을 변경하도록 하겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/AdminViewController.swift:
```diff
@@ -48,4 +51,20 @@ class AdminViewController: UIViewController {
+                }
+        })
+    
+        pieChartView.segments = segments
```
-  여기 코드들은 대부분 pie graph를 그리기 위한 코드니까 PieGraphView에 있어야 할 것 같습니다.
- 이전까지 뷰 객체에 데이터를 한꺼번에 전달하지 않아야 한다고 강조했습니다. segments를 넘기지 않고 PieGraphView가 DataSource 프로토콜을 채택한 모델 객체에게 물어보는 구조로 만드세요.

AppleCEO replied:
- changeSegments 메소드를 PieGraphView 로 이동하였습니다.
- PieGraphView 가 pie Chart 를 그릴 때 DataSource 에 음료수 별로 재고를 물어보도록 구성해봤습니다.

godrm commented on VendingMachineApp/VendingMachineApp/DrinkMenu.swift:
```diff
@@ -46,4 +46,15 @@ enum DrinkMenu: Int, CaseIterable {
+        case .top: return "티오피"
+        case .hot6: return "핫식스"
+        case .pepsiCoke: return "펩시콜라"
+        }
```
클래스를 만들었으면 클래스 속성으로 해결할 수 있는건 별도 매칭하는 함수를 만들지 마세요.
클래스 인스턴스와 화면 요소를 매핑하는 것만 하면 객체(혹은 클래스)에 모든 값이 들어있는 구조가 좋습니다. 
그렇지 않으면 타입이 바뀌거나 타입이 추가되기만 해도 여기 메소드들을 모두 수정해야 합니다.

AppleCEO replied:
getImageName 과 getKoreaName 메소드를 하나로 합쳤습니다. 
enum 이 너무 커지지 않게 클래스 속성으로 최대한 모든 값이 들어있도록 구성하겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,276 @@
+        formatter.maximumFractionDigits = 1
+        return formatter
+    }()
+}
```
여기 있는 타입들 중에서는  UIKit과 직접적으로 관련이 없는 코드들이 있습니다. 
파일을 구분할 때도 관련있는 것끼리 묶어놓는게 좋습니다. UIKit과 관련이 없는 코드를 여기에 다 묶어놓을 필요는 없습니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,276 @@
+    /// An array of structs representing the segments of the pie chart.
+    var segments = [LabelledSegment]() {
+        // Re-draw view when the values get set.
+        didSet { setNeedsDisplay() }
```
setNeedsDisplay()를 호출하고 있긴하네요. 모든 속성을 뷰 객체 외부에서 바꾸고 바꿀때마다 다시 그리는 구조를 하고 있습니다. 
이 값이 얼마나 될지 모르는데 모두 가지고 있는것은 좋지 않습니다. 
꼭 이렇게 구현해야만 할까요? 다시 그리는 것을 최소화할 수는 없을까요? 
@IBInspectable 을 활용한 접근 방식은 좋지만, 그리는 방법을 최소화하는 최적화를 고민해보세요.
`https://developer.apple.com/videos/play/wwdc2018/220/` 자료를 공부해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,276 @@
+                var segmentCenter = viewCenter
+                if segments.count > 1 {
+                    segmentCenter = segmentCenter
+                        .projected(by: radius * textPositionOffset, angle: halfAngle)
```
들여쓰기 단계가 깊어졌네요. 4단계 이내로 개선해볼 수 있으면 좋겠습니다.

AppleCEO replied:
들여쓰기 단계 1단계로 개선했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,276 @@
+    
+    /// The value of the segment.
+    var value: CGFloat
+}
```
이 객체는 뷰에 데이터를 전달해주기 위한 작은 ViewModel 인가요? 다른 파일로 분리해보세요.

AppleCEO replied:
PieGraphView 에 같이 있던 SegmentLabelFormatter 와 LabelledSegment 구조체들을 각각 파일을 생성해주었습니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,276 @@
+    @IBInspectable
+    var textPositionOffset: CGFloat = 0.67 {
+        didSet { setNeedsDisplay() }
+    }
```
@IBInspectable 을 활용한 방식은 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Printable.swift:
```diff
@@ -20,3 +20,7 @@ protocol SellDrinkPrintable {
     func printSellDrinkList(handler: (Int, DrinkMenu)->())
 }
+
+protocol pieGraphDrawable {
```
프로토콜 이름도 타입 이름이니까 대문자로 사용하세요.

godrm replied:
이것도 수정이 됐나요?

AppleCEO replied:
놓쳤었네요. 지금 수정했습니다.

godrm reviewed (commented):
뷰 객체를 그리기 위한 DataSource 객체에 대한 피드백입니다. 프로토콜을 채택하고 있지만 DataSource 기준으로 관계를 개선해보세요.

AppleCEO committed:
추가 - 피드백을 받고 super 메소드 호출을 추가

AppleCEO committed:
수정 - 글자가 커서 차트를 넘어가는 경우가 발생해 글자 크기 줄임

AppleCEO committed:
수정 - 내부 메소드를 실행하기 때문에 의미없이 반복되는 매개변수 vendingMachine 제거

AppleCEO committed:
수정 - 의미없는 공백 제거

AppleCEO committed:
수정 - draw 를 하지 않음에도 메소드명에 있던 애매한 부분 수정

AppleCEO committed:
이동 - changeSegments 메소드를 pieChartView 객체로 이동

godrm commented on VendingMachineApp/VendingMachineApp/AdminViewController.swift:
```diff
@@ -37,28 +38,28 @@ class AdminViewController: UIViewController {
+        pieChartView.segmentLabelFont = .systemFont(ofSize: 15)
         
-        stock.drawPieGraph(handler:
+        vendingMachine.drawPieGraph(handler:
```
이렇게 생각하세요. 
- View 는 화면에 그리기 위한 시각적인 정보를 처리하고, 요청하고, 반복해서 그리는 역할을 한다. 
- ViewController는 그리는 동작을 직접 하지는 않고, 어떤 뷰가 그려야 할지 아닐지, 어떤 뷰의 데이터는 누가 갔고 있는지 연결해주고 중간에서 관리만 합니다. 
- 그런 의미에서 colors 데이터는 누가 갖고 있어야 할까요? 
- pieChartView 속성을 바꾸면 다시 그려진다는 것은 그 데이터를 뷰가 아닌 다른 객체가 바꿔야 하기 때문에 명시적이지 않습니다. 그러면 뷰는 수동적인 구조가 됩니다. 더 유연한 구조는 뷰가 능동적으로 필요한 데이터를 DataSource에게 물어서 가져가는 방식입니다. 
- 그럼 DataSource 는 (지금까지는) VendingMachine 이어야 한다고 생각했지만, 자판기의 모든 데이터가 필요한게 아니라 View에게 필요한 정보만 간추려서 전달하면 됩니다. 그런 역할을 담당하는 별도 객체여도 됩니다. 하지만 ViewController가 그 역할을 하는 것은 좋지 않습니다.
- 데이터 흐름이 ViewController -> View 로 가는게 아니라, View -> DataSource 로 가야한다는 게 중요합니다. 
- draw를 직접 호출하지는 않지만, ViewController가 할 일은 setNeedsDisplay()를 호출해주는 겁니다. 다시 그리라고 하면 필요하면 View가 데이터를 갖고 있는 DataSource에게 물어서 그리면 되는겁니다. 
그래서 이 함수 구현 부분을 개선해보라고 했던 겁니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -272,4 +272,17 @@ class PieGraphView : UIView {
+        
+        segmentLabelFont = .systemFont(ofSize: 15)
+        
+        vendingMachine.drawPieGraph(handler:
```
PieGraph 에서 VendingMachineManagementable 전체 인터페이스가 필요한건가요?
아니면 출력을 하기 위한 인터페이스 (프로토콜)만 필요한건가요? 
PieGraphView는 자판기에서만 쓸 수 있는 뷰인건가요? 
피드백만을 지나가려고 수정하지 말고 의미있는 코드를 만들기 위해서 피드백을 기준으로 수정했으면 좋겠습니다.

AppleCEO committed:
수정 - pieGraphDrawable 을 채택하면 addSegment 메소드를 수행할 수 있도록 구성

메소드 동작도 음료수 종류에 따라 카운트를 segment 로 추가하도록 변경

AppleCEO committed:
수정 - 모델이 dataSource 에 값을 물어보도록 수정

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -211,6 +213,16 @@ class PieGraphView : UIView {
+                    if count > 0 {
+                        segments.append(LabelledSegment(color: colors[drink.rawValue], name: drink.getKoreaName(), value: CGFloat(count)))
+                    }
+            }, drinkMenu: drink)
```
거의 다 된거 같은데... DataSource가 필요한 segments를 미리 자판기와 연계해서 갖고 있어야 합니다. 이름이 데이터 소스인데 View가 값을 넣어주고 있네요!! 
당연히 View에서 데이터를 갖고 있는 데이터 소스에 필요한 데이터를 물어봐야 하는거죠. 
뷰가 segments를 갖고 있을 필요가 없습니다. 더구나 여기 draw() 메소드는 초당 60번 - 120번씩 불리는 함수입니다. 
매번 저걸 계산하고 있을 필요가 없습니다. 데이터 소스가 갖고 있는걸 가져와서 그리기만 하면 됩니다. 
segments 속성이 없이 데이터 소스에 요청해서 그리도록 개선해보세요. 
그 과정이 OutputView가 모델에 요청해서 출력하는 흐름과 동일한 흐름이 됩니다.

AppleCEO replied:
view 에서 데이터 소스에 필요한 데이터를 물어보면 데이터 소스에서 view 에 배열로 돌려줘도 될까요?

godrm replied:
배열 전체를 넘겨주는 것보다는 메소드를 분리하면 어떨까요? 
그릴게 몇 개가 있어? 1번째는 뭘 그릴까? 2번째는 뭘 그릴까? 이런 식으로요...

AppleCEO replied:
segments 속성을 제거했습니다. 
view 가 필요한 데이터들을 각각 model 에 물어보도록 구성해봤습니다.

AppleCEO committed:
제거 - DrinkMenu enum 이 너무 커지고 있어서 이름을 리턴하는 메소드 하나로 통합

AppleCEO committed:
추가 - 그래프를 그리는데 필요한 메소드 추가

getSegmentsValueTotal : 각 세그먼트의 값들의 총합(음료수 총합) getSegemntsCount : 세그먼트의 개수(음료수 종류 수) drawSegment : 세그먼트를 그리는 메소드

AppleCEO committed:
수정 - 세그먼트의 배열을 모델에서 한번에 받아오지 않고 필요한 정보를 하나씩 물어보도록 수정

기존의 segments.count 를 getSegmentsCount 메서드로 물어보고 각 요소들 값들의 총합인 valueCount 를 getSegmentsValueTotal 로 물어보고 세그먼트의 숫자만큼 drawSegment 에 index 를 보내서 호출하여 세그먼트를 그리도록 구성

AppleCEO committed:
수정 - depth 를 줄이기 위해서 filter 함수로 변경

AppleCEO committed:
제거 - 테스트하는데 사용한 소스 제거

AppleCEO committed:
수정 - 프로토콜명 첫글자를 대문자로 수정

AppleCEO committed:
생성 - pie graph 를 그리는데 필요한 구조체들을 별도 파일로 이동

godrm commented:
filter 로 바꿔도 되는 비슷한 코드가 몇 개 더 있었는데 혹시 변경이 됐나요?

AppleCEO commented:
네 고차함수로 가능한 부분들은 전부 수정했습니다.

godrm commented:
수고하셨습니다.
